### PR TITLE
(TWILL-133) Make the ZKClient stoppable even it is still trying to connect

### PR DESCRIPTION
**Pull request by chtyim**
- Also include refactoring of the DefaultZKClient class to simplify
  shutdown logic

### CODE REVIEW

File: twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java [hunk 1]
```diff
@@ -420,23 +444,18 @@ protected void terminated() {
...
-        } catch (InterruptedException e) {
-          notifyFailed(e);
-        } finally {
-          eventExecutor.shutdown();
-          executorStopped = true;
-
-          // If the executor state is terminated, meaning the terminate() method is triggered,
-          // call notifyStopped() if it hasn't been called yet.
-          if (eventExecutor.isTerminated() && stopNotified.compareAndSet(false, true)) {
+      eventExecutor.submit(new Runnable() {
```
Comment (hsaputra) [inline twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java]: Why we need to run stop on separate thread?

File: twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java [hunk 2]
```diff
@@ -479,12 +498,26 @@ public void run() {
...
+    /**
+     * Closes the given {@link ZooKeeper} if it is not null. If there is InterruptedException,
+     * it will get logged.
+     */
+    private void closeZooKeeper(@Nullable ZooKeeper zk) {
+      try {
+        if (zk != null) {
+          zk.close();
+        }
+      } catch (InterruptedException e) {
```
Comment (hsaputra) [inline twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java]: Should we reset the interrupt flag by calling Thread.currentThread().interrupt() here ?

**Comment by hsaputra** [comment]
Did a pass and have couple questions.

File: twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java (hunk 1 above)
Comment (chtyim) [inline twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java]: It is needed. Added comments to explain the possible race. Also, guava AbstractService expects stop execute asynchronously.

File: twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java (hunk 2 above)
Comment (chtyim) [inline twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/DefaultZKClientService.java]: added

**Comment by hsaputra** [comment]
+1
LGTM

**Comment by chtyim** [comment]
Find a bug in LeaderElection that can causes unit-test runs unstable. Checked-in in the last commit.

**Comment by hsaputra** [comment]
Good catch. We should squash the commits to help preserve the history then re +1

**Pull request closed without merging**